cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -594,10 +766,18 @@ private module SsaInput implements SsaImplCommon::InputSig {
    */
   predicate variableWrite(IRBlock bb, int i, SourceVariable v, boolean certain) {
     DataFlowImplCommon::forceCachingInSameStage() and
-    variableWriteCand(bb, i, v) and
+    (
+      variableWriteCand(bb, i, v) or
+      sourceVariableIsGlobal(v, _, _, _)
```

Comment:
Why is this change needed?

Reply:
Oh, that's a good point. I should add a comment to this. The reason is as follows:

`variableWriteCand` restricts the set of writes in the second SSA phase to those writes that target variables that are determined to be live by the first SSA phase. However, the first SSA phase doesn't know about global variables (as you can see, nothing in this PR touches the `ssa0/SsaInternals.qll` file). So it may be the case that a write to a global variable is determined to not be live by the first SSA phase (because there's no synthetic "final use of the global variable" at the end of the function body).

Now, there are two ways of handling this situation:
1. Teach `ssa0/SsaInternals.qll` about global variables (i.e., by doing the same transformation this PR does, but also do it in the `ssa0/SsaInternals.qll` file), or
2. Exclude global variables from the pruning by saying "either the variable isn't pruned away in the first SSA phase, or it's a global variable (like we do in this PR).

The reason I chose the second approach is that I don't expect the first phase to prune much (if anything) away in the case of global variables (since, if you write to a global variable then there's a very very high chance that you also read it _somewhere_ in the database). So it wouldn't really prune anything away, and it would only increase the amount of data the first SSA phase is working with.

Does that make sense?

Reply:
Ok, so that I mostly got from other part of the code. What confuses me here is that we both have this addition and the following just a bit lower:
```
    or
    exists(GlobalDef global |
      global.hasIndexInBlock(bb, i, v) and
      certain = true
    )
```
Why isn't it sufficient to just have this latter bit?

Reply:
Oh! I think you're right 😂. Let me double check that I can remove that `sourceVariableIsGlobal` disjunct.

Reply:
Oh, wait. No, I think it's actually needed. Basically, there are two kinds of writes to global variables:
1. An actual write to the variable in the body of the function. That's this part:
```ql
exists(DefImpl def | def.hasIndexInBlock(bb, i, v) |
  if def.isCertain() then certain = true else certain = false
)
```
2. A synthesizes initial definition of the global variable. That's this part:
```ql
exists(GlobalDef global |
  global.hasIndexInBlock(bb, i, v) and
  certain = true
)
```

Consider a program like:
```cpp
void write() {
  global = 42;
}

void read() {
  use(global);
}
```

the write to `global` is an instance of `DefImpl`, and so hits the first case. But we don't want to prune away that write (since there is a read of the variable. Just not in the same callable).

Reply:
Thanks for the explanation. That clarifies things.